### PR TITLE
Implement submodules key from v2 config

### LINKED ISSUE #4464: Add submodule configuration
Migrated from https://github.com/rtfd/readthedocs-build/issues/30

Seems a number of projects are hitting issues with submodule clone steps, and this should be configurable in our config. Because git doesn't provide the mechanism for optional submodules, we can do something like:

```yaml
submodules:
  include: []
  recursive: false
```

as implicit default. And:

```yaml
submodules:
  include:
    - foo/
    - bar/
```

As include example.

```yaml
submodules:
  exclude:
    - foo/
```

^ is the exclude example.

Legacy projects implemented before this change would have an implicit configuration of:

```yaml
submodules:
  include: all
  recursive: true
```

It's not possible to include and exclude submodules, this is a validation error.

This would eventually alter the code that rtfd/readthedocs.org uses for submodule cloning, and would only selectively clone things? The changes there are likely moving to specific submodule clone commands.

- [ ] Currently submodule cloning takes place in the git checkout step. This will need to be moved to a new step after we parse the configuration
- [ ] New behavior will be to not clone submodules automatically
- [ ] If possible, we won't recursively clone as well. Maybe this is another option on the spec?
- [ ] We'll add a new `project.Feature`, plus migration, to lock old users into the old behavior (all submodules are cloned)
- [ ] We can perhaps also detect that a project has submodules that won't be cloned and raise a warning. This warning shouldn't be a build failure, but a notification on-site/through email?

### PR DESCRIPTION
**Pull request by stsewd**
This PR on top of #4456. It should be merged after. New changes are from 2fee268 

Closes #4464

### CODE REVIEW

**Comment by stsewd** [comment]
The only thing missing is to test v1 with submodules, but I kind of need what I have in #4394

**Comment by humitos** [comment]
I will wait until the base PR gets merged to review this.

**Comment by agjohnson** [comment]
This should be ready for rebase. It's already set to master base.

File: readthedocs/rtd_tests/tests/test_config_integration.py [hunk 1]
```diff
@@ -729,43 +729,6 @@ def test_sphinx_configuration(
         append_conf.assert_called_once()
         move.assert_called_once()
 
-    @patch('readthedocs.doc_builder.backends.sphinx.BaseSphinx.move')
-    @patch('readthedocs.doc_builder.backends.sphinx.BaseSphinx.append_conf')
-    @patch('readthedocs.doc_builder.backends.sphinx.BaseSphinx.run')
-    def test_sphinx_fail_on_warning(
```
Comment (stsewd) [inline readthedocs/rtd_tests/tests/test_config_integration.py]: Not sure how this landed here... Anyway this test was added in  https://github.com/rtfd/readthedocs.org/pull/4456/files#diff-5a2cc4147895068b33fe5bf89e4e6e08R685

File: readthedocs/projects/tasks.py [hunk 2]
```diff
@@ -443,6 +439,22 @@ def run_setup(self, record=True):
 
         return True
 
+    def additional_vcs_operations(self):
+        # TODO: execute with lock
```
Comment (stsewd) [inline readthedocs/projects/tasks.py]: I think this step should be done with the context management, the one that creates a lock

File: readthedocs/vcs_support/backends/git.py [hunk 3]
```diff
@@ -53,38 +55,55 @@ def set_remote_url(self, url):
...
+        submodules_in_config = (
+            config.submodules.exclude != ALL or
+            config.submodules.include
+        )
+        if (self.project.has_feature(Feature.SKIP_SUBMODULES) or
+                not submodules_in_config):
             return False
+
+        # Keep compatibility with previous projects
         code, out, _ = self.run('git', 'submodule', 'status', record=False)
```
Comment (stsewd) [inline readthedocs/vcs_support/backends/git.py]: We were running this code to check for submodules previously, so we need to keep the same behavior here

File: readthedocs/vcs_support/base.py [hunk 4]
```diff
@@ -43,6 +43,7 @@ class BaseVCS(object):
 
     supports_tags = False  # Whether this VCS supports tags or not.
     supports_branches = False  # Whether this VCS supports branches or not.
+    supports_submodules = False
```
Comment (stsewd) [inline readthedocs/vcs_support/base.py]: This way we can say what vcs needs to do an extra step

**Comment by agjohnson** [comment]
Is this still WIP? Just needs a rebase if not

**Comment by stsewd** [comment]
I only have a doubt in https://github.com/rtfd/readthedocs.org/pull/4493#discussion_r213903992, and then test this for real with the sphinx or mkdocs implementation

**Comment by stsewd** [comment]
I figure out that we don't need the lock in the submodules operation, I guess the lock is there so we don't end with several builds of the same version. This PR is tested and ready for review.

**Comment by ax3l** [comment]
@agjohnson with this PR merged (thx a lot! :) ), is there anything we can add to our `.readthedocs.yml` already that will exclude recursive git submodule init?

**Comment by stsewd** [comment]
@ax3l this isn't deployed yet, and it's part of the v2 of our configuration file, I'm trying to get some users to a beta soon, please let your project here https://github.com/rtfd/readthedocs.org/issues/1621#issuecomment-425954082

**Pull request merged**